waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,381 @@
+            [clojure.data.codec.base64 :as b64]
+            [clojure.java.io :as io]
+            [clojure.string :as string]
+            [comb.template :as template]
+            [metrics.counters :as counters]
+            [plumbing.core :as pc]
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :as codec]
+            [ring.util.response :as response]
+            [waiter.auth.authentication :as auth]
+            [waiter.middleware :as middleware]
+            [waiter.util.date-utils :as du]
+            [waiter.util.utils :as utils]
+            [waiter.metrics :as metrics])
+  (:import (java.io ByteArrayInputStream ByteArrayOutputStream)
+           (java.nio.charset Charset)
+           (java.security.cert CertificateFactory)
+           (java.util.zip Deflater DeflaterOutputStream)
+           (javax.xml.parsers DocumentBuilderFactory)
+           (org.opensaml Configuration DefaultBootstrap)
+           (org.opensaml.xml.security.x509 BasicX509Credential)
+           (org.opensaml.xml.signature SignatureValidator)
+           (org.opensaml.xml.validation ValidationException)))
+
+(def charset-format (Charset/forName "UTF-8"))
+
+(defn str->bytes
+  "Get bytes of a string"
+  [some-string]
+  (.getBytes some-string charset-format))
+
+(defn bytes->str
+  "Get string from bytes"
+  [some-bytes]
+  (String. some-bytes charset-format))
+
+(defn deflate-bytes
+  "Gzip compress bytes"
+  [str-bytes]
+  (let [out (ByteArrayOutputStream.)
+        deflater (DeflaterOutputStream.
+                   out
+                   (Deflater. -1 true) 1024)]
+    (.write deflater str-bytes)
+    (.close deflater)
+    (.toByteArray out)))
+
+(defn str->deflate->base64
+  "Gzip compress bytes and base 64 encode"
+  [deflatable-str]
+  (let [byte-str (str->bytes deflatable-str)]
+    (bytes->str (b64/encode (deflate-bytes byte-str)))))
+
+(defn base64->str
+  "Decode base 64 string"
+  [string]
+  (let [byte-str (str->bytes string)]
+    (bytes->str (b64/decode byte-str))))
+
+(defn get-idp-redirect
+  "Return Ring response for HTTP 302 redirect."
+  [idp-url saml-request relay-state]
+  (response/redirect
+    (str idp-url
+         "?"
+         (let [saml-request (str->deflate->base64 saml-request)]
+           (codec/form-encode
+             {:SAMLRequest saml-request :RelayState relay-state})))))
+
+(defn validate-saml-assertion-signature
+  "Checks that the SAML assertion has a valid signature."
+  [assertion saml-signature-validator]
+  (if-let [signature (.getSignature assertion)]
+    (try
+      (.validate saml-signature-validator signature)
+      (catch ValidationException ex
+        (throw (ex-info "Could not authenticate user. Invalid SAML assertion signature."
+                        {:status 400} ex))))
+    (throw (ex-info "Could not authenticate user. SAML assertion is not signed."
+                    {:status 400}))))
+
+(defn create-document-builder
+  "Create new xml document builder"
```

Comment:
Creates the document builder for what purpose?

Reply:
so you can build an xml document with it